File: services/core-service/src/main/java/com/itsumori/beneaththepine/core/task/application/TaskRepository.java
package com.itsumori.beneaththepine.core.task.application;

import com.itsumori.beneaththepine.core.task.domain.NextAction;
import com.itsumori.beneaththepine.core.task.domain.Task;
import com.itsumori.beneaththepine.core.task.domain.TaskStatus;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface TaskRepository {
    void createConfirmedAction(Task task, NextAction nextAction);

    Optional<Task> findByIdForUser(UUID taskId, UUID userId);

    List<Task> listForUser(UUID userId, Optional<TaskStatus> status, int limit);

    Task update(Task task);
}

File: services/core-service/src/task/application/task.repository.ts
import type { ConfirmedAction, Task, TaskStatus } from '../domain/task';

export abstract class TaskRepository {
  abstract createConfirmedAction(task: Task, action: ConfirmedAction): Promise<void>;
  abstract findForUser(id: string, userId: string): Promise<Task | undefined>;
  abstract listForUser(userId: string, status: TaskStatus | undefined, limit: number): Promise<Task[]>;
  // Lock and read the current owned row before applying the domain transition.
  abstract changeForUser(id: string, userId: string, change: (task: Task) => Task): Promise<Task>;
}
